File: sort/java/SelectionSort.java
// selection sort repeatedly selects the smallest or largest element from the unsorted portion of the list
// and moves it to the sorted portion of the list
// TC: O(n2)
// Mem: O(1) temp
public class SelectionSort {
    public void sort(int[] arr) {
        if (arr == null) {
            throw new IllegalArgumentException();
        }

        int n = arr.length;
        // move starting pointer
        for (int i = 0; i < n - 1; i++) {
            int min_idx = i;
            // unsorted array
            for (int j = i; j < n; j++) {
                if (arr[j] < arr[min_idx]) {
                    min_idx = j;
                }
            }
            // swap
            int temp = arr[min_idx];
            arr[min_idx] = arr[i];
            arr[i] = temp;
        }

    }

    public static void main(String[] args) {
        SelectionSort ss = new SelectionSort();
        int[] arr = new int[] { 64, 25, 12, 22, 11 };
        ss.sort(arr);

        for (int i = 0; i < arr.length; i++)
            System.out.println(arr[i]);
    }
}

File: sort/go/selection_sort.go
package main

// func main() {
// 	nums := []int{0, 9, 3, 5, 7, 9, 2, 3}
// 	selectionSort(nums)
// 	fmt.Println(nums)
// }

func selectionSort(nums []int) {
	// last idx : 2
	// i=0, j=1,
	// i=0, j=2

	// i=1, j=2
	for i := 0; i < len(nums)-1; i++ {
		minIdx := i
		for j := i + 1; j < len(nums); j++ {
			if nums[minIdx] > nums[j] {
				minIdx = j
			}

		}
		temp := nums[minIdx]
		nums[minIdx] = nums[i]
		nums[i] = temp
	}
}
